add unsupervised calculation on gains for root locus plot.

gonmolina opened the pull request:
Add gains up to a tolerance is achived.
@slivingston I wait for reviews. I have a file with some examples that are an exersice taken from Franklyn book. Let me know if you need it to test the algorithm. Do I have to put in a particular place? Where? Thanks in advance.

slivingston commented:
@gonmolina can you add a citation in the docstring to the book that you used?

slivingston commented:
I am sorry for my delay; I lost track of my GitHub notifications feed. I anticipate finishing reviewing this within 2 days.

gonmolina commented:
I added a book as a reference. 
I added possibility to draw a grid on s-plane similar to matlab (argument grid = True).
I have tested a lot this function and works fine. 
I have wrote a script with several tests with a lot of cases and root locus figures seem to be fine. If you think this script can help you to test the algorithm I can add it to the repository (please tell me where is the right place to add it). 
I have added support of 0 degrees root locus (positive feedback or  k<0) in the algorithm. This works fine in all cases that I have tested, but I couldn't solve when the num.order=den.order. To solve this problem a complete new solution have to be written, mainly when lines are plotted and when roots are sorted. I raise an error when a root locus of this kind is demanded. 

 @slivingston I wait for your review.
Thanks in advance.
Gonzalo

slivingston commented:
@gonmolina thanks working on it now.

slivingston commented on control/rlocus.py:
```diff
@@ -127,14 +137,137 @@ def root_locus(sys, kvect=None, xlim=None, ylim=None, plotstr='-', Plot=True,
+
+
+def _k_max(num, den, real_break_points, k_break_points):
+    """" Calculation the maximum gain for the root locus shown in tne figure"""
```
misprint: "tne" should be "the". Also, the first word should some verb. E.g., "Calculate"

slivingston commented on control/rlocus.py:
```diff
@@ -157,29 +290,33 @@ def _systopoly1d(sys):
         curpoly = denp + k * nump
         curroots = curpoly.r
+        if len(curroots) < denp.order:
+            curroots = np.insert(curroots, len(curroots), np.inf)  # if i have less poles than open loop is because  i
```
The comment here overflows into the next line. For ease of reading, it should be moved to the line before. This style for block comments is given [in PEP 8](https://www.python.org/dev/peps/pep-0008/#block-comments). E.g., after I edit the comment text a little,
```python
if len(curroots) < denp.order:
    # If I have fewer poles than open loop, it is because I
    # have one at infinity.
    curroots = np.insert(curroots, len(curroots), np.inf)
```

slivingston commented:
@gonmolina thanks. The changes are good.

slivingston commented:
Actually considering the failures on CI testing, there may have been a regression. I am trying to reproduce locally.

gonmolina commented:
Is there any way to do the test locally in my machine without doing a push? 
I think I can fix it....

slivingston commented:
There are several new changes outside the scope of the previous review, so I will review these when you tell me that they are ready.

Several of the commit messages mention fixing of bugs. If these existed before this PR, can you add regression tests for them, i.e., tests that demonstrate the bugs?

gonmolina commented:
I made 2 changes:
- I added line 268 to ensure kmax> 0 when all the singular points are in 0. Problems getting kvect when deal with systems of the form 1/s^n. I do not think this is the problem while building.
- I greatly increased the gain after the kmax in line 209 so that the lines of the  roots locus reach the zeros (in the figure) for all cases. I think this can be the problem since several groups of equal roots turn out to be about the same for too large gains. Solution: fewer points after kmax, smaller maximum gain, and separate them logarithmically. This problem was clear when I was trying to solve 0 degrees root locus with for systems with equal number of zeros and poles.

A regression test for the first problem is:

import control as ctrl
sys1=ctrl.tf([1],[1,0,0])
r,k = ctrl.rlocus(sys1, grid=True)

slivingston commented:
@gonmolina it looks like you are permitting me to push changes to the branch associated with this pull request. Can I do so, to add the regression test that [you suggest above](https://github.com/python-control/python-control/pull/138#issuecomment-312406694)?

gonmolina commented:
yes, you can do it.

murrayrm commented:
@slivingston I'm trying to clean up some of the PRs over the holiday break.  Let me know if you are still working on this one.

murrayrm commented:
I've run Travis CI on a local branch and all tests are OK: [results](https://travis-ci.org/murrayrm/python-control/builds/324465228).

@gonmolina, @slivingston Are we ready to merge or is more testing required?  Seems like this has to be better than what was there before (placeholder).
